waiter/src/waiter/token.clj

```diff
@@ -165,23 +194,31 @@
               (kv/delete kv-store owner-key)))
           ; Delete owner map
           (kv/delete kv-store token-owners-key))
-        (let [owner->tokens (->> tokens
-                                 (map (fn [token] (let [{:strs [owner]} (kv/fetch kv-store token)]
-                                                    {:owner owner
-                                                     :token token})))
-                                 (filter :owner)
-                                 (group-by :owner))
-              owner->owner-key (->> owner->tokens
+        (let [owner->index-entries (->> tokens
+                                        set
+                                        (map
+                                          (fn [token]
+                                            (let [{:strs [deleted owner] :as kv-data} (kv/fetch kv-store token)
+                                                  token-etag (kv-data->etag kv-data)]
+                                              (cond-> {:deleted (true? deleted)
+                                                       :owner owner
+                                                       :token token}
+                                                      token-etag (assoc :etag token-etag)))))
+                                        (filter :owner)
+                                        (group-by :owner))
+              owner->owner-key (->> owner->index-entries
                                     keys
                                     (map (fn [owner] [owner (new-owner-key)]))
                                     (into {}))]
           ; Create new owner map
           (kv/store kv-store token-owners-key owner->owner-key)
           ; Write each owner node
-          (doseq [[owner tokens] owner->tokens]
+          (doseq [[owner index-entries] owner->index-entries]
             (let [owner-key (get owner->owner-key owner)
-                  token-set (->> tokens (map :token) set)]
-              (kv/store kv-store owner-key token-set))))))))
+                  index-entries (->> index-entries
+                                     (map #(select-keys % [:etag :token]))
```

Comment:
Do we need to select `:deleted` as well?

Reply:
Yes.

Reply:
Done.